Basic Instrument View functionality with PyVista

jclarkeSTFC opened the pull request:
### Description of work
This contains the new Instrument View with some basic functionality. You can pick detectors, view their spectra, and view their info. You can change TOF and contour range limits for the colouring of the views. In this version the 3D view is in the top section, while the lower section will show one of the 2D projections. Side-by-side projection hasn't been implemented yet, this is work for a future PR.

To view the new Instrument View there are two methods:

- Right-click on a workspace in Workbench and select `Show Instrument (New)`
- Run `python -m instrumentview --file <your_file>`

The restriction for `cmake>4` has been relaxed to allow `>=3.21` since on Windows it's not possible to resolve the `mantid-developer` environment with `cmake>=4`. Once dependencies are compatible then `conda` will use version 4 automatically.

On macOS, `libboost` version 1.88 is not compatible with the dependencies on this branch.

#### Purpose of work

Fixes #38364 
Fixes #38424 
Fixes #38422 
Fixes #38421 
Fixes #38427 
Fixes #38423 

This is part of the Instrument View Overhaul epic.

### To test:

Right-clicking on a detector will select it. 

Test the package builds and a dev build because we've added a dependency, and the new instrument view itself is a new Python module.

Package builds: https://builds.mantidproject.org/job/Core_Team_test_pipeline/276/

The `qt-main` pin has been relaxed from `5.15.8` to `5.15.*` so the packages need a good test, not just the new instrument view.

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Instrument View 2.0 using pyvista

jclarkeSTFC committed:
Initial work on projections

Spherical and cylindrical, using matplotlib

jclarkeSTFC committed:
Move code into new widget, make MVP

jclarkeSTFC committed:
Plot projection in lower view

jclarkeSTFC committed:
Edit contour range using text boxes

jclarkeSTFC committed:
Add pyvista and pyvistaqt to developer environment

jclarkeSTFC committed:
Implement time-of-flight filtering

jclarkeSTFC committed:
Add projection selection combo box

jclarkeSTFC committed:
Add new Instrument View to workspace context menu

Also added a couple of tests. One existing test was clearing the ADS for
the whole test class, which means the tests would pass/fail depending on
the order they ran, so I've fixed that as well.

jclarkeSTFC committed:
Add many tests and fix a couple of problems

jclarkeSTFC committed:
Update year on copyright notice

jclarkeSTFC committed:
Add tests for projections

jclarkeSTFC committed:
Remove hard-coded file path from module

The file to load is now an argument when launching the instrumentview
module. It also now uses the more general Load algorithm, instead of
the separate algorithm, e.g. LoadEventNexus

jclarkeSTFC committed:
Host detector plot inside the main widget.

jclarkeSTFC committed:
Add multi-select

New toggle will allow user to draw a rectangle and select all detectors
within that rectangle. Drawing the rectangle or not is toggled with
the 'r' key (set by PyVista).

jclarkeSTFC committed:
Set icon when launching module

When running the instrumentview module separately to Workbench, set
the application icon to something Mantid.

jclarkeSTFC committed:
Improve detector info when multiple selected

The QTextEdit handles longer text better than a QLineEdit, now we'll
get a scroll wheel when the text from the selected detector(s) is too
long for one line.

jclarkeSTFC committed:
Refactor projection classes and add tests

jclarkeSTFC committed:
Add doc strings and some tidying up.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jclarkeSTFC committed:
Cppcheck fixes

jclarkeSTFC committed:
Inline function only used in one place.

jclarkeSTFC committed:
Attempt to glean information from CI test failures

jclarkeSTFC committed:
Workbench needs instrumentview as a dependency

jclarkeSTFC commented:
Points from meeting with Richard and Tom:
- Performance improvement for WISH
- Individually select multiple detectors
- Change x unit
- Synchronise picked detectors in projection and 3D views
- Line plot toolbar
- Potential rearrangement of view, e.g. tabs for projection and 3D, extend line plot horizontally
- Detector point size when zooming
- Problem loading WISH file from system test data

sf1919 commented:
Is there any way to split this up? It's a giant PR as it is and if there is more to do....

jclarkeSTFC committed:
Some performance improvements to instrumentview

A couple of calculations can be done in parallel, and some of the
calculations can be done in a more efficient way.

jclarkeSTFC committed:
Speed up getting detector integrated counts

Do it in C++ instead of Python, this makes it much faster. Also, keep
track of the detector IDs, this is useful, and not intuitive how to
do.

jclarkeSTFC committed:
Ensure instrumentview is built before workbench

GuiMacielPereira committed:
Add basic multiplt point picking to main plotter

- Allows user to select and deselect multiple detectors
- Model stores visibility of picked detectors

jclarkeSTFC committed:
Tidy up dependencies

joblib is already in meta.yaml, so no need to add it again, plus it's
no longer needed for instrumentview. Add a dependency of instrumentview
on mantidqt in the CMakeLists.txt file to ensure things build in the
correct order.

GuiMacielPereira committed:
Correct rectange selection

- Rectangle selection now selects enclosed points
- Works in the same way as point picking, if point is picked it unpicks
the point and vice-versa

GuiMacielPereira committed:
Basic picking sync with projections plotter for point picking

- Point picking working for both plots and is synchronised
- Slow when many detectors are picked, due to plotting many detectors
everytime
- Retangle picking not working

jclarkeSTFC committed:
Fix tests following recent changes

Ignore the tests for drawing empty instruments because they take a long
time to run, and currently they do no work on Linux. However they are
useful for testing manually.

GuiMacielPereira committed:
Renamed variables and added model public methods

jclarkeSTFC committed:
Change to use Instrument 2.0

Instead of looping through the components in the instrument, we now use
the information on the DetectorInfo class on the workspace to get IDs
and position, etc. This will hopefully make things simpler and
easier to maintain in the future. This also eases the difficulties
working out how to map components to spectra, since it's easy with
detector IDs.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jclarkeSTFC committed:
Tidy up unused C++ exports

These methods were added to draw detector shapes, but are no longer
used, since we're not currently drawing detector shapes.

GuiMacielPereira committed:
Remove detector info from projection inputs

The projections depend only on sample, root and detector positions, so I
made that clearer by setting these as the inputs to the class.
Changed model getter method for projection positions to be consistent
with main positions.

GuiMacielPereira committed:
Store picked detectors in list instead of dict

Makes it more consistent with the other lists stored in the model

GuiMacielPereira committed:
Changed variable names and arg types

GuiMacielPereira committed:
Fix presenter unit tests

GuiMacielPereira committed:
Fix projection unit test

GuiMacielPereira committed:
Fix cylindrical projection unit tests

GuiMacielPereira committed:
Fix sphere projection unit test

GuiMacielPereira committed:
Vectorize calculation of projections

Using numpy arrays instead of parallelising calculation is about 1000X
faster (I checked the times with a MARI workspace).
Also makes code cleaner.

GuiMacielPereira committed:
Vectorise x correction in projection

Correction for the x range after projection now is 10X faster by using
numpy arrays instead of for loops.
Running time measured with MARI.
Also a bit cleaner to read.

GuiMacielPereira committed:
Small cleanups of picking functions

sf1919 commented:
This is becoming a huge PR - is there any way of splitting it up?

jclarkeSTFC committed:
Remove references to deprecated arg and property

GuiMacielPereira committed:
Refactor to use arrays instead of Detector class

This refactor does not use a Detector class but instead keeps track of
detector properties in several arrays with the same ordering and the
same size.

The main advantage of this design is that it avoids creating multiple
maps between detector IDs and other properties and allows for easily
selecting non-monitor or monitor detectors by using the boolean mask:

eg. self._detector_ids[~self._is_monitor] selects the non-monitor
detector ids

There are some small (probably negligible) performance improvements, due
to the fact that all required arrays are calculated at the beggining and
there is less need to constantly iterate over a list[Detector] to
retrieve lists of properties.
Since previously a lot of the getter methods in the model were already
retrieving lists, I think it makes more sense for the model to keep
track of these lists/arrays instead of keeping track of many different
Detector() objects.

Finally, Mantid's design of DetectorInfo is the same: DetectorInfo acts
as a big map where we can retrieve properties of detectors through
indexing. I think it might be awkward to use DetectorInfo inside a
Detector class, because technically DetectorInfo contains information on
all detectors, so I am not sure it should be contained whithin each
detector.

GuiMacielPereira committed:
Updated unit tests after refactor

- Updated a few unit tests
- Deleted one unit test because I changed the function in a way that
will require rewriting it

sf1919 commented:
Ignore the OSX failure - the Windows one seems genuine.

GuiMacielPereira committed:
Clean picking a little bit

jclarkeSTFC committed:
Handle detectors with no corresponing spectrum

On some instruments there are detectors listed in the DectectorInfo
class that do not have a corresponding spectrum in the workspace. Work
around this by using a different method to map detector IDs to
workspace indices. It already existed, I just had to add a Python
export for it. Also added a test for this situation.

jclarkeSTFC commented:
Test packages for Linux [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/250/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/250/)

jclarkeSTFC committed:
Workbench needs pyvista and pyvistaqt

There's currently no instrumentview conda package, so we need to
specify pyvista and pyvistaqt as workbench dependencies.

jclarkeSTFC commented:
Test package Linux [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/252/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/252/)

jclarkeSTFC committed:
Add more model tests

GuiMacielPereira committed:
Fix picking inaccuracy in projection plotter

The trick was to decrease the tolerance argument in
enable_point_picking(), the default value is 0.025 and decreasing it to
0.01 yields an accurate picking experience

jclarkeSTFC committed:
Fix projection tests after refactor

jclarkeSTFC committed:
Remove minor version pin for qt-main

The pyvista dependency requires a newer version of qt-main than 5.15.8

jclarkeSTFC commented:
Test package Linux [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/253/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/253/)

jclarkeSTFC committed:
Button for clearing detector selection.

jclarkeSTFC committed:
Return counts as a double

This maintains consistency with the existing code for integrating
spectra.

jclarkeSTFC committed:
Remove reference to side-by-side projection

This will be implemented in a future PR.

jclarkeSTFC commented:
Build packages all platforms: [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/254/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/254/)

jclarkeSTFC commented:
Looks like a build problem on ARM, I'll mark this as draft again and investigate

jclarkeSTFC committed:
Pin pyqt5-sip to <12.13

When building packages on ARM, without this pin conda picks up
pyqt5-sip 12.17, then we get an error saying that 'sip-build: '12.15'
is not a supported ABI version', since our version of pyqt5 requires
an earlier version.

jclarkeSTFC committed:
Don't do certain things if off-screen rendering

Attempt to fix mysterious GL error on the Windows agents, hopefully
by avoiding calling these methods we can avoid the problem.

jclarkeSTFC committed:
Separate tests for the view

Mock out the view in the workspace widget tests, then separately test
everything in the view in a new test class.

jhaigh0 commented on qt/python/instrumentview/instrumentview/test/test_model.py:
```diff
@@ -0,0 +1,147 @@
+        self.assertEqual(min(integrated_spectra), model._data_min)
+        self.assertEqual(max(integrated_spectra), model._data_max)
+
+    # TODO: Add test for picked_detectors_info_text
+    # TODO: Add tests for other picking methods
```
are these to be done in this pr?

jhaigh0 reviewed (changes requested):
Most of the feedback is stylistic. The new instrument view code is well-structured and relatively easy to follow. A few extra things:

 - Should the new instrument view already be in the workspace context menu given it's not yet finished (are we going to get curious people using the nightly and complaining about it not having all the old features etc.)
 - I'm unsure if the instrument view presenter should be constructed by the view. I think in the past, I've had the view have a method called 'subscribe_presenter' or something like that. I think the idea is that you can swap and change views / models. It's also very useful for testing (can subscribe a mock presenter). Same goes for model constructed by the presenter.
 - I'd have stylistic feedback about the interface itself, but I think this maybe isn't the time for that?
 - There is a lot of type hinting which is nice but maybe 10-20% of function aren't type hinted / have incomplete hinting. I'm not going to let that stop me approving the pr but just thought I'd mention it.

GuiMacielPereira committed:
Initialise presenter with model and view

Added view method subscribe_presenter to be called from inside
presenter.
Presenter is now initialised with a view and a model

GuiMacielPereira committed:
Isolate connect statements in single method

Since presenter is subscribed to the view, it makes more sense that
connections are forcibly established after the presenter is subscribed.
This way there is no chance of a presenter slot being called when the
presenter does not exist.

GuiMacielPereira committed:
Fix unit tests

Updated unit tests so that they pass

GuiMacielPereira committed:
Add missing unit tests for model

GuiMacielPereira committed:
Use dataclass for DetectorInfo

GuiMacielPereira committed:
Remove unused methods from ComponentInfo and CSGObject

GuiMacielPereira committed:
Use transform instead of for loop

Co-authored-by: Jonathan Haigh <Jonathan.Haigh@stfc.ac.uk>

GuiMacielPereira committed:
Update cppcheck suppressions

GuiMacielPereira committed:
Rename methods for consistency

Co-authored-by: Jonathan Haigh <Jonathan.Haigh@stfc.ac.uk>

GuiMacielPereira committed:
Rename classes for projections

Co-authored-by: Jonathan Haigh <Jonathan.Haigh@stfc.ac.uk>

GuiMacielPereira committed:
Remove redundant __init__()

Co-authored-by: Jonathan Haigh <Jonathan.Haigh@stfc.ac.uk>

GuiMacielPereira committed:
Update qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py

Co-authored-by: Jonathan Haigh <35813666+jhaigh0@users.noreply.github.com>

GuiMacielPereira commented on qt/python/instrumentview/instrumentview/Projections/cylindrical_projection.py:
```diff
@@ -0,0 +1,37 @@
+    ):
+        super().__init__(sample_position, root_position, detector_positions, axis)
+
+    def _calculate_2d_coordinates(self, detector_position: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
```
Unfortunately this is only present in 3.12 and we're still using 3.11

sf1919 commented:
Genuine Linux, OSX and Windows test failures for commit [5e5a1a3](https://github.com/mantidproject/mantid/pull/39081/commits/5e5a1a313f346efa288036c8b6be38f7ec4f1299). Tests will re-run with new commit

GuiMacielPereira committed:
Revert "Update qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py"

This reverts commit 5e5a1a313f346efa288036c8b6be38f7ec4f1299.

GuiMacielPereira committed:
Rename variables and delete unused comments

Co-authored-by: Jonathan Haigh
<35813666+jhaigh0@users.noreply.github.com>

GuiMacielPereira committed:
Clearer description for --file argument

Co-authored-by: Jonathan Haigh <Jonathan.Haigh@stfc.ac.uk>

GuiMacielPereira committed:
Add tests for projections

Co-authored-by: Jonathan Haigh <Jonathan.Haigh@stfc.ac.uk>

GuiMacielPereira committed:
Correct argument name

GuiMacielPereira committed:
Add type hints

Co-authored-by: Jonathan Haigh <Jonathan.Haigh@stfc.ac.uk>

GuiMacielPereira committed:
Prevent GUI from opening when detectors or instrument not present

In the case where there are no detectors present, there is no point in
opening instrument view.

GuiMacielPereira committed:
Fix opening instrumentview from mantid

Had forgotten to update the trigger to opening from mantid

GuiMacielPereira committed:
Rename menu option to Experimental

GuiMacielPereira commented on qt/python/instrumentview/setup.py:
```diff
@@ -0,0 +1,19 @@
+
+setup(
+    name="instrumentview",
+    install_requires=["mantidqt"],
```
Created https://github.com/mantidproject/mantid/issues/39689 so we remember this

sf1919 commented:
The OSX failure on commit [ffc3eb2](https://github.com/mantidproject/mantid/pull/39081/commits/ffc3eb2a06d4a22ff4a72866310c08ea47bf8212) appears to be for pyvista so will need fixing.

GuiMacielPereira committed:
Separate view from presenter further

- Renamed methods to getters and setters
- Took parsing of min and max from edit boxes to presenter

GuiMacielPereira committed:
Update unit tests

GuiMacielPereira commented on qt/python/instrumentview/instrumentview/FullInstrumentViewWindow.py:
```diff
@@ -0,0 +1,325 @@
+        self._contour_range_min_edit.textEdited.connect(self._on_contour_limits_updated)
+        self._contour_range_max_edit.textEdited.connect(self._on_contour_limits_updated)
+        self._tof_min_edit.textEdited.connect(self._on_tof_limits_updated)
+        self._tof_max_edit.textEdited.connect(self._on_tof_limits_updated)
```
For future reference, I spoke with @jhaigh0 and we agreed that keeping the connections in the view allows for more flexibility when switching views. This is because it allows the presenter not to have knowledge about the view that is currently in use, so the presenter always remains unchanged regardless of the view chosen.

GuiMacielPereira committed:
Remove development comment

jhaigh0 reviewed (dismissed):
Go on then

jclarkeSTFC committed:
Revert min cmake to previous, keep 4 available

cmake>=4 causes dependency resolution problems on Windows and macOS.
By removing the <4 pin instead of setting >=4 we will start using
cmake 4 when dependencies are available. There's no particular reason
why we need to enforce cmake>=4.

jclarkeSTFC committed:
Merge branch 'main' into pyvista_test

jclarkeSTFC committed:
Reduce boost version to 1.86 on macOS

Boost 1.88 is not compatible with the other dependencies on this branch
on macOS. Also I consolidated the boost pin variable in
conda_build_config since both boost libraries seem to be on the same
release schedule, I don't think we want to mix and match (not sure that
would even work)

jclarkeSTFC commented:
Package build all platforms: [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/276/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/276/)

jclarkeSTFC committed:
Merge branch 'main' into pyvista_test

jclarkeSTFC committed:
Use same picking tolerance for both windows

Co-authored-by: Gui Pereira

jclarkeSTFC committed:
Fix view test following argument change

jhaigh0 reviewed (dismissed):
The dependency pinning changes are needed to make the environment resolve given the new dependencies. They can be looked at after this is merged.

RichardWaiteSTFC reviewed (changes requested):
Done some functional testing on SXD workspace and a subset of SXD. Overall feels responsive once loaded and the UI look and feel is nice an clean.
Just a few issues (some of which may only apply to me?) and also some features which may already be planned/done in a separate PR (apologies if that's the case) - but obviously happy if any of them go in here.

Potential issues for this PR:

- [ ] Took ~30s to open SXD on windows and mantid hung – should it be on a separate thread?
- [ ] Default window size bit small for me (on my screen resolution)
- [ ] Right clicking on 2D projection adds spectra to mini-plot even when Select Multiple Detectors = False (only works 3D view – rectangular select – could be icon?)
- [ ] To translate 3D view need to press scroll button, on old UI was right-click (perhaps easier for laptop users with no mouse?)
- [ ] When zooming in on peak in 2D plot the marker size gets very small so mostly white space which makes it difficult to see which pixels contribute to a peak
- [ ] When min/max TOF or colorscale limits are invalid (e.g. max < min) no user prompt or resetting of values
- [ ] Should get default 2D projection from IDF if specified (e.g. spherical Z for SXD)
- [ ] Seems to be chopping of some SXD detectors (banks 1-3) in Spherical Z?

New features needed (but maybe not in this PR, and maybe not in MVP?):

- [ ] Want to left click draw a rectangle to zoom into region of 2D plot (atm have to scroll and translate as zoom is always around center of widget)
- [ ] Need to be able to zoom into miniplot – e.g. to find d-spacing of peak
- [ ] Home button or equivalent to reset view in 3D and 2D plots
- [ ] Need ADS observer, if workspace is replaced reopen UI, if deleted close – atm doesn’t crash just prints warnings
- [ ] Colorbars on RHS of plots probably makes more sense for most screen aspect ratios
- [ ] Slider to resize panel on left
- [ ] Could 2D markers be square (on old instrument view I think it actually used the size of the component)
- [ ] Instrument tree viewer to select specific banks (maybe this wasn’t in MVP?)
- [ ] Want to be able to undock 1D plot
- [ ] Want to export 1D plots to workspace – trivial once thing for selecting units
- [ ] Want option not to fix aspect ratio of 2D and 3D plots – 
- [ ] Summing of selected spectra (planned in unit selection PR)

jclarkeSTFC commented:
Thanks Richard, I've linked issues for the items that we're going to do separately. Anything without an issue we'll either do it in this this PR or at least discover that it's hard enough to do as a separate issue, in which case I'll update this list.

   * [ ]  Took ~30s to open SXD on windows and mantid hung – should it be on a separate thread?

   * [ ]  Default window size bit small for me (on my screen resolution)

   * [ ]  Right clicking on 2D projection adds spectra to mini-plot even when Select Multiple Detectors = False (only works 3D view – rectangular select – could be icon?)

   * [ ]  To translate 3D view need to press scroll button, on old UI was right-click (perhaps easier for laptop users with no mouse?)

   * [ ]  When zooming in on peak in 2D plot the marker size gets very small so mostly white space which makes it difficult to see which pixels contribute to a peak

   * [ ]  When min/max TOF or colorscale limits are invalid (e.g. max < min) no user prompt or resetting of values

   * [ ]  Should get default 2D projection from IDF if specified (e.g. spherical Z for SXD) #39746 

   * [ ]  Seems to be chopping of some SXD detectors (banks 1-3) in Spherical Z?


New features needed (but maybe not in this PR, and maybe not in MVP?):

   * [ ]  Want to left click draw a rectangle to zoom into region of 2D plot (atm have to scroll and translate as zoom is always around center of widget) #39747 

   * [ ]  Need to be able to zoom into miniplot – e.g. to find d-spacing of peak #39748 

   * [ ]  Home button or equivalent to reset view in 3D and 2D plots #39749 

   * [ ]  Need ADS observer, if workspace is replaced reopen UI, if deleted close – atm doesn’t crash just prints warnings

   * [ ]  Colorbars on RHS of plots probably makes more sense for most screen aspect ratios

   * [ ]  Slider to resize panel on left

   * [ ]  Could 2D markers be square (on old instrument view I think it actually used the size of the component)

   * [ ]  Instrument tree viewer to select specific banks (maybe this wasn’t in MVP?) #39750 

   * [ ]  Want to be able to undock 1D plot #38425 

   * [ ]  Want to export 1D plots to workspace – trivial once thing for selecting units #38429 

   * [ ]  Want option not to fix aspect ratio of 2D and 3D plots – #39752 

   * [ ]  Summing of selected spectra (planned in unit selection PR) #38429

GuiMacielPereira committed:
Clean x gap correction

- Set default period to 2 pi
- Clean part to find biggest gap in x range by using numpy detectors

The only change in functunality is that if there are two or more gaps
with the exact same size, the gap chosen might or might not be the same
as before. I don't believe this is a problem, since this is not common
and the point of the routine is to avoid the biggest gap in the x range.
If two such big gaps are present, choosing one over the other should not
improve the visualisation of the detectors one way or the other.

GuiMacielPereira committed:
Refactor x gap correction further and fix unit tests

- Removed an if statement because with the new way of finding the
biggest gap, x_from will always be smaller than x_to
- The consequence is that some instruments might have their projections
mirrored in the new instrument view

GuiMacielPereira committed:
Change background to black and increase point size

The black background and bigger points helps with identifying peaks in
the instrument

GuiMacielPereira committed:
Update unit tests

GuiMacielPereira committed:
Fix initial window size

GuiMacielPereira committed:
Update checkbox label and add button for clearing projection

GuiMacielPereira committed:
Merge branch 'main' into pyvista_test

GuiMacielPereira committed:
Merge branch 'address-review-sugestions' into pyvista_test

GuiMacielPereira committed:
Merge branch 'main' into pyvista_test

GuiMacielPereira committed:
Add double slider to tof and contour ranges

jclarkeSTFC committed:
Add superqt dependency to mantidworkbench

jclarkeSTFC committed:
Pin gdk-pixbuf=2.42.12 on Windows

On Windows there is a packaging error when calling
Scripts\.gdk-pixbuf-post-link.bat that is referencing a path that
doesn't exist. Pin to the specific build variant of 2.42.12 that
worked previously.

GuiMacielPereira committed:
Merge branch 'add-double-slider' into pyvista_test

GuiMacielPereira committed:
Setup model in a separate QThread

Due to long loading times when first opening instrument view, this
commit opens the view with a loading gif to let the user know that the
interface is being set-up.

GuiMacielPereira committed:
Open instrument view with default projection

Each instrument has a default projection which the user can specify in
the IDF file. This commit exposes the default projection to python and
uses it when first opening instrument view window

GuiMacielPereira committed:
Add workspace observers for clear, rename and delete

If a workspace is renamed, the model is updated to hold the new
workspace.
If the workspace is deleted or ADS is cleared, instrument view is
closed.
The observers are disconnected each time the main window is closed, as
not to trigger involuntary function calls.

jclarkeSTFC committed:
Fix ADS behaviour on closing the instrument view

Also, run callbacks on a QAppThreadCall to avoid hanging, seems to be
a problem on Windows especially.

GuiMacielPereira committed:
Fix failing unit tests

GuiMacielPereira committed:
Set up QThread using QObject instead of inheriting

Instead of creating a QThread by inheriting QThread (runs on the main
thread), create a QObject that runns on a QThread, which ensures that
worker gets run in a separate thread.

From the C++ side, need to release the GIL so that the python
interpreter can run that section in parallel rather than switching back
and forth between main and separate thread (default behaviour in
python).

GuiMacielPereira committed:
Fix updating of tof values

Previously moving the tof range caused the limits of the color bar to
change.
This is undesirable as there is already a slider for controling the
colorbar.
Now the tof values update only the counts in the figures

GuiMacielPereira committed:
Add splitter to allow user to resize panels

Added two splitters: one for resizing in the horizontal direction
(between the options column and the plotters) and another splitter for
resizing in the vertical direction (between options and line plot)

GuiMacielPereira committed:
Remove PythonInterfaces from Framework

This reverts a previous change that was causing build problems.
In principle, the Framework should not depend on interfaces.
This is not a huge deal, since the speed up from releasing the GIL was
not substancial, and the mantid hand still happened on openning.

GuiMacielPereira committed:
Remove threading from presenter

Opening a separate QThread was causing issues in
test_workbenchwidget.py.
The current thread work was not actually being run in parallel due to
the GIL, so removing this for now will not impact performance.

GuiMacielPereira committed:
Merge branch 'mantidproject:main' into pyvista_test

thomashampson committed:
pin conda to avoid post-link.bat error

GuiMacielPereira committed:
Merge branch 'main' into pyvista_test

RichardWaiteSTFC commented on qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py:
```diff
@@ -0,0 +1,151 @@
+        self._source_position = np.array(self._component_info.sourcePosition()) if has_source else np.array([0, 0, 0])
+        self._detector_ids = np.array(self._detector_info.detectorIDs())
+        self._detector_positions = np.array([self._detector_info.position(i) for i in range(len(self._detector_ids))])
+        self._spherical_positions = np.array([self._detector_info.position(i).getSpherical() for i in range(len(self._detector_ids))])
```
Is it possible (maybe in a different PR) to keep `self._detector_positions` as a V3D in the line above, then call `.getSpherical()` before converting to numpy array (avoid getting it twice) - might be a bottleneck?

RichardWaiteSTFC replied:
Also consider using `np.asarray`  - I think this sometimes avoids copying when making the array - not sure if this will be possible for V3D?

RichardWaiteSTFC replied:
The other alternative for now to get a quick speed up is to call `CreateDetectorTable` algorithm (with new-ish option `IncludeDetectorPosition=True`) then you can get the spherical  like so `tbl.column('Position')` - which returns a list of V3Ds on which you can call `getSpherical`

RichardWaiteSTFC reviewed (changes requested):
Thanks for all the comments that have been addressed. 
Viewing empty WISH data does seem to crash Mantid on windows with following error
```
Traceback (most recent call last):
File "C:\MantidUnstableInstall\bin\Lib\site-packages\superqt\sliders\_generic_slider.py", line 301, in paintEvent
opt = self._styleOption
^^^^^^^^^^^^^^^^^
File "C:\MantidUnstableInstall\bin\Lib\site-packages\superqt\sliders\_generic_slider.py", line 374, in _styleOption
self.initStyleOption(opt)
File "C:\MantidUnstableInstall\bin\Lib\site-packages\superqt\sliders\_generic_slider.py", line 217, in initStyleOption
option.tickInterval = self._to_qinteger_space(self._tickInterval)
^^^^^^^^^^^^^^^^^^^
TypeError: 'float' object cannot be interpreted as an integer
```
The workspace  is plotted though so might be OK if handled?

GuiMacielPereira committed:
Fix crash on empty instruments

Empty instruments were attempting to set a range of the double slider to
(0, 0), which causes a crash.
Tried to catch the error with try/except but for some reason the
exception is not caught.
So opted for checking the limits before calling slider.setRange()

GuiMacielPereira committed:
Merge branch 'main' into pyvista_test

GuiMacielPereira committed:
Merge branch 'mantidproject:main' into pyvista_test

jhaigh0 reviewed (changes requested):
Found some things:
1. drag the vertical splitter all the way to the left so that it 'squishes' the side with the plot, got an error reporter pop-up that I could continue from. I think this kind of problem has been fixed in other workspaces.
2. Noticed some strange behaviour when editing (i.e doubling it) the data while instrument view is open. At the very least it doesn't update, but sometimes there are errors. I think `InstrumentVewADSObserver` should patch into the `replaceHandle` method.
3. Typically, when you edit the max box for 'Contour Range' and press tab, if it's above the max value it resets to the max. If you fill the box with digits (i.e add 0 to the end until you reach max characters) then press tab, the value isn't updated.
4. In the preview plot, I think the menu for choosing lin/log x/y is strange? There's three sub-menus called linear x, log x, symlog x, which each have a further sub-menu of linear y, log y, symlog y. I don't seem to be able to set the x axis to log, plus if I set y to log then back to linear, the plot is a bit zoomed in. In general the plot is very slow to update to settings changes.
5. Is there a way to use the rectangular selection on the projection widget? I can make it work on the 3D view, on the projection it zooms in. Also, on the projection widget, if you left-click (like a user / me might do on first use to see what it does) and move your mouse even a single pixel, it will zoom into a tiny area and probably just give you a black screen. It would be nice if this was less sensitive.
6. When testing with POLARIS00098534, I couldn't select detectors in the projections wiget (could in the 3d one).
7. When typing into the min/max boxes, if you enter a max value less than the min value (or vice versa) the box correctly sets itself to be equal to the min value, but the colours on the projections don't update.


Of these issues, I think 1. and 4. are the most significant. 2. should be fixed (especially considering live data) but I think it's likely to be a bit of work, so happy for it to come later. 
In general, I'm happy for someone else to decided what needs to be done in this pr, I'm still not too clear and how ready for users this needs to be.

Oh and last thing, not sure why the package is versioned 6.14.8?

GuiMacielPereira commented:
I have written up these issues in #39882, #39883, #39884, #39885, #39886 and #39887, so that these changes can go in in separate PRs.

As for the strange version of the package, that's just because I typed a random version (I tried doing proper naming in the past but go versioning errors late in the pipeline so I did not want to risk it for this one)

sf1919 commented:
This PR is getting larger and larger. Given it is new functionality and clearly marked as experimental in the menu I think we need to merge this in now. It has passed all it's CI checks and empty instrument crash highlighted by @RichardWaiteSTFC . Meanwhile bug raised by @jhaigh0 have been written up as separate issues. It will only be available in the nightly until release so there is some time to get working on the most important bug fixes.

I feel it would be quicker to deal with the bugs as separate PRs rather than continue working on this one.
